### PR TITLE
Fix address association on users with no addresses

### PR DESCRIPTION
**Pull request by mbj**

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb This is ready for review.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 1]
```diff
@@ -17,12 +17,21 @@ class CheckoutController < Spree::StoreController
...
     before_action :check_authorization
 
-    before_action :setup_for_current_state, only: %i[edit]
-
     helper 'spree/orders'
 
     rescue_from Spree::Core::GatewayError, :with => :rescue_from_spree_gateway_error
 
+    def edit
+      case @order.state
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Why not move this case statement into `setup_for_current_state`? Obviously `setup_for_current_state` will need to be changed a bit, but otherwise I think the logic would still apply.

Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: If you don't like the use of `before_action` I would still move this case statement into a private method and explicitly call it from here instead of inlining it.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 2]
```diff
@@ -17,12 +17,21 @@ class CheckoutController < Spree::StoreController
...
+      when 'address'
+        edit_address_setup
+      when 'payment'
+        edit_payment_setup
+      end
+
+      render :edit
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: I believe you could replace this with `super` and get the same results.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 3]
```diff
@@ -108,22 +117,18 @@ def completion_route
...
-        method_name = :"before_#{@order.state}"
-        send(method_name) if respond_to?(method_name, true)
-      end
-
-      def before_address
+      def edit_address_setup
         user = try_spree_current_user
+
         if user
           @order.associate_user(user)
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Given that the `else` condition was removed I would rewrite this line as:

```ruby
@order.associate_user(user) if user
```

**Comment by dkubb** [comment]
@mbj I left a few comments in this PR.

FYI, are the modified methods mutation tested?

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 1 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: `setup_for_current_state` actually only was called for `:edit` action, hence using a hook for causing a sequential effect was wrong, because using a standard ruby sequence of statements gives the same effect at less complexity.

I'm fine to move out the setup into a private method, but I wounder why we should extract it when there is no other call side? Is it too long already?

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 3 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: Yeah better.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 1 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: @dkubb The `render :edit` call does not seem to be needed (super implementation does not exist). Factoring out the state setup just to have it called as the only body of the action does seem to be overdoing it. So I left it in-place.

**Comment by mbj** [comment]
> BTW, are the modified methods mutation tested?

Now: Yes. And I could remove all sorts of dead code. Its also integration tested against our main app.

**Comment by mbj** [comment]
@dkubb Ready for the next pass.

File: core/lib/spree/core/controller_helpers/order.rb [hunk 4]
```diff
@@ -7,8 +7,6 @@ module Order
         PG_LOCK_NOT_AVAILABLE = 'PG::LockNotAvailable:'.freeze
 
         included do
-          before_filter :set_current_order
```
Comment (dkubb) [inline core/lib/spree/core/controller_helpers/order.rb]: Won't this cause issues with being removed, or is the current order set elsewhere?

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 5]
```diff
@@ -108,27 +115,6 @@ def completion_route
...
-        user = try_spree_current_user
-        if user
-          @order.associate_user(user)
-        else
-          @order.bill_address ||= Address.build_default
-          @order.ship_address ||= Address.build_default if @order.checkout_steps.include?('delivery')
-        end
-      end
-
-      def before_payment
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Should this removal be in this commit? It seems kind of unrelated to the spirit of the change.

File: core/lib/spree/core/controller_helpers/order.rb (hunk 4 above)
Comment (mbj) [inline core/lib/spree/core/controller_helpers/order.rb]: If this is paired with the changes in https://github.com/MountainRoseHerbs/spree_auth_devise/pull/13 the only call side that benefited from it will get fixed.

The call was entirely redundant, and also was the source of various duplicate SQL queries, and inferenced with caching the wrong state in the ivars (For example when the user is just logged in, but the ivar is already resolved against the guest token, not factoring in the `user_id`).

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 5 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: This removal contains the several layers of redundant indirection and conditional deep nested call side that was lifted directly to `#edit`.

You'll notice that `before_address` is without many conditionals now directly in `#edit`.

Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: And the removed layers of indirection that could hit `before_payment` where found when lifting `before_edit` contents.

Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: > You'll notice that before_address is without many conditionals now directly in #edit.

I noticed that, but I didn't see this functionality moved anywhere else. What I would've liked to see, as a reviewer, would be for it to be in it's own commit and an explanation why it's no longer necessary to set `@payment_sources`.

Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: `@payment_sources` was found to be dead the moment I mutation covered the call side that delegated to `before_address` or `before_payment`. Support for user specific payment sources was removed earlier (existing credit cards).

**Comment by mbj** [comment]
@dkubb Ready for the next pass
